raft/raft.go

```diff
@@ -120,8 +120,15 @@ func (pr *Progress) waitSet(w int)       { pr.Wait = w }
 func (pr *Progress) waitReset()          { pr.Wait = 0 }
 func (pr *Progress) isUnreachable() bool { return pr.Unreachable }
 func (pr *Progress) reachable()          { pr.Unreachable = false }
-func (pr *Progress) unreachable()        { pr.Unreachable = true }
-func (pr *Progress) shouldWait() bool    { return (pr.Unreachable || pr.Match == 0) && pr.Wait > 0 }
+
+func (pr *Progress) unreachable() {
+	pr.Unreachable = true
+	// If the remote becomes unreachable, there is big probability that at least
+	// one MsgApp fails to be sent out. Fall back to bad path to recover it.
+	pr.Next = pr.Match + 1
```

Comment:
this is not quite right i guess. we should only decrease next to pr.match+1 if we were on the optimistic sending path. or it will decrease next to 1 (since match is 0). 

```
if pr.Match != 0 {
    pr.Next = pr.Match + 1
}
```